### PR TITLE
Test for issue #15408

### PR DESCRIPTION
**Pull request by barche**
Also remove some stale instances of `jl_new_closure` and fix a compile error.

### CODE REVIEW

File: src/codegen.cpp [hunk 1]
```diff
@@ -1298,7 +1298,7 @@ void jl_extern_c(jl_function_t *f, jl_value_t *rt, jl_value_t *argt, char *name)
             assert(!imaging_mode);
 #if defined(USE_MCJIT) || defined(USE_ORCJIT)
             jl_ExecutionEngine->addGlobalMapping(llvmf->getName(),
-                    (void*)getAddressForOrCompileFunction(llvmf));
```
Comment (yuyichao) [inline src/codegen.cpp]: Why is this removed?

File: test/core.jl [hunk 2]
```diff
@@ -690,6 +690,18 @@ let A = [1]
...
 end
 
+# Pointer finalizer (issue #15408)
+let A = [1]
+    function f(to_finalize::Array{Int64,1})
+      to_finalize[1] = 0
+      nothing
+    end
+    cf = cfunction(f, Void, (Array{Int64,1},))
+    finalizer(A, cf)
```
Comment (yuyichao) [inline test/core.jl]: Passing a `cfunction` to `finalizer` is not allowed. It shouldn't cause a big issue in this case but it is better to use a real c function (possibly defined in `libccalltest`) instead.

Comment (yuyichao) [inline test/core.jl]: (in general, calling any julia runtime in a function pointer finalizer is not allowed.)

File: src/codegen.cpp (hunk 1 above)
Comment (barche) [inline src/codegen.cpp]: It gave me an invalid conversion from void* to uint64_t on arch linux

Comment (yuyichao) [inline src/codegen.cpp]: Seems that the `JuliaOJIT` and `llvm::ExecutionEngine` doesn't agree on this API. I guess we should probably fix `JuliaOJIT::addGlobalMapping`.

File: test/core.jl (hunk 2 above)
Comment (barche) [inline test/core.jl]: I didn't know about `libccalltest`, it would be much more appropriate for this, I'll rewrite the test.

Comment (barche) [inline test/core.jl]: The C finalizer should take a `jl_value_t*` as argument, so I would need to include julia.h from ccalltest.c. Would that be acceptable?

Comment (yuyichao) [inline test/core.jl]: It's acceptable but not necessary, you can just use a `void*` instead. Also make sure to NOT call any `jl_`* functions in the finalizer...

File: src/codegen.cpp [hunk 3]
```diff
@@ -1299,9 +1299,12 @@ void jl_extern_c(jl_function_t *f, jl_value_t *rt, jl_value_t *argt, char *name)
         } else {
             // Otherwise we use a global mapping
             assert(!imaging_mode);
-#if defined(USE_MCJIT) || defined(USE_ORCJIT)
+#if defined(USE_ORCJIT)
             jl_ExecutionEngine->addGlobalMapping(llvmf->getName(),
                     (void*)getAddressForOrCompileFunction(llvmf));
+#elif defined(USE_MCJIT)
+            jl_ExecutionEngine->addGlobalMapping(llvmf->getName(),
+                    getAddressForOrCompileFunction(llvmf));
```
Comment (yuyichao) [inline src/codegen.cpp]: It's probably better to actually match the signature but this is probably good enough for now =)

File: test/core.jl [hunk 4]
```diff
@@ -690,6 +690,16 @@ let A = [1]
     @test x == 1
 end
 
+# Pointer finalizer (issue #15408)
+let A = [1]
+    testlib = Libdl.dlopen("libccalltest")
+    ccall(Libdl.dlsym(testlib, "set_c_int"), Void, (Cint,), Cint(1))
```
Comment (yuyichao) [inline test/core.jl]: `ccall((:set_c_int, "libccalltest"))` and `cglobal((:finalizer_cptr, "libccalltest"))` See the [ccall test](https://github.com/JuliaLang/julia/blob/4b85d85a9f7256718bbba7a45fc68fb454dd0532/test/ccall.jl). This test could probably be moved there too.

Comment (yuyichao) [inline test/core.jl]: The `Cint` constructor calls shouldn't be necessary.

**Comment by yuyichao** [comment]
Note that updating the commit in a PR does not generate any notification so it is better to leave a message when you've made change and the PR is ready for review again.

File: src/ccalltest.c [hunk 5]
```diff
@@ -362,6 +363,21 @@ JL_DLLEXPORT struct_big test_big(struct_big a) {
     return a;
 }
 
+JL_DLLEXPORT int get_c_int(void)
+{
+  return c_int;
```
Comment (tkelman) [inline src/ccalltest.c]: 4-space indent

**Comment by barche** [comment]
OK, thanks for the detailed review. I'm sure it's perfect now ;)

I don't know anything about `jl_ExecutionEngine` so I don't feel comfortable doing anything more about that, but maybe we should create a new issue for that?

**Comment by yuyichao** [comment]
Lgtm
